### ISSUE TITLE
Word2vec - CBOW, Skip-gram

### ISSUE THREAD

**Issue by sophryu99** [sophryu99/TIL#14]
# Word2vec - CBOW, Skip-gram

Word Embedding: 비정형화된 text를 숫자로 바꿔줌으로써 사람의 언어를 컴퓨터의 언어로 번역하는 것으로, 단어의 의미는 **벡터의 형태**로 표현된다.

<br/>

## Word2Vec

**Word2Vec**: 단어를 **벡터화** 해주는 프레임워크

- 단어간의 유사도를 계산하다는 점에 있어 원-핫 벡터의 한계를 보완한다.

```markdown
고양이 + 애교 = 강아지
한국 - 서울 + 도쿄 = 일본
박찬호 - 야구 + 축구 = 호나우두
```

- 위의 예시 처럼 단어간의 벡터 값에 대한 연산을 통해 유사도를 확인할 수 있다.

<br/>

### CBOW - Continous Bag of words

**주변의 단어**로 **가운데 들어가는 단어를 예측**하는 방법이다. 

<img width="400" alt="Screen Shot 2021-06-02 at 12 32 10 PM" src="https://user-images.githubusercontent.com/46921003/120420066-8e9f8480-c39e-11eb-823a-79daf76b6edc.png">

#### Logic

1. 중심 단어 전후로 **몇 개의 단어를 참고**할지를 설정한다. (설정 범위를 window라고 함)

2. 윈도우가 계속 움직이며 중심 단어, 주변 단어를 바꾸어 가며 학습한다. (**sliding window**)

   <img width="474" alt="Screen Shot 2021-06-02 at 12 47 15 PM" src="https://user-images.githubusercontent.com/46921003/120421117-ac6de900-c3a0-11eb-953a-38a39d04187c.png">

3. 입력층에 **사용자가 설정한 윈도우 크기**와 해당 범위 안의 **주변 단어들이 원-핫 벡터**로 들어간다.

4. 입력층과 출력층 사이에 projection layer가 한 층만 존재하며, 활성화 함수가 존재하지 않는다.

   1. 활성화 함수가 존재하지 않는 대신 룩업 테이블이 존재하며, projection layer(투사층)는 일반적인 은닉층과는 다르다. 
   2. 아래 그림에서는 각 주변 단어의 원-핫 벡터를 x로 표기하였다. 입력 벡터는 원-핫 벡터이다. i번째 인덱스에 1이라는 값을 가지고 그 외의 0의 값을 가지는 입력 벡터와 가중치 W 행렬의 곱은 **사실 W행렬의 i번째 행을 그대로 읽어오는 것**과(lookup) 동일하다. 그래서 이 작업을 **룩업 테이블(lookup table)**이라고 부른다. 

   <img width="400" alt="Screen Shot 2021-06-02 at 12 55 24 PM" src="https://user-images.githubusercontent.com/46921003/120421737-cd830980-c3a1-11eb-8592-5a05b3a4c8e2.png">

5. **입력층과 projection layer**의 **가중치(W1) = V X M** 행렬

6. **Projection Layer와 출력층**의 **가중치(W2) = M X V** 행렬 

   1. V: 단어 집합의 크기
   2. M: Projection Layer의 크기 -> 임베딩 하고 난 후의 벡터의 차원이 됨

7. 입력층의 벡터들이 Projection Layer에서 모여, **이것들의 평균**을 구한다. (아래의 그림)

8. <img width="400" alt="Screen Shot 2021-06-02 at 1 07 59 PM" src="https://user-images.githubusercontent.com/46921003/120422599-8f86e500-c3a3-11eb-9108-bc17a472d6bd.png">

9. CBOW는 softmax 함수로 벡터값을 최종 산출한다. 

   <img width="400" alt="Screen Shot 2021-06-02 at 1 12 33 PM" src="https://user-images.githubusercontent.com/46921003/120422850-323f6380-c3a4-11eb-8367-7fb6560db295.png">

   - 이렇게 구해진 평균 벡터는 두번째 가중치 행렬 W2와 곱해진다. 곱셈의 결과로는 원-핫 벡터들과 차원이 V로 동일한 벡터가 나온다. Softmax 함수로 인한 출력값은 0과 1 사이의 실수로, 각 원소의 총 합은 1이 되는 상태로 바뀐다. 

10. 손실 함수로는 아래와 같은 cross-entropy를 사용한다. y hat이 y를 **정확히 예측한 경우**의 cross-entropy 값은 **0**이 되기 때문에 **해당 값을 최소화** 하는 것이 중요하다.

    <img width="219" alt="Screen Shot 2021-06-02 at 1 18 21 PM" src="https://user-images.githubusercontent.com/46921003/120423240-01abf980-c3a5-11eb-82b2-fa62c18b2722.png">

11. 이러한 구조를 기반으로 역전파를 수행하며 가중치 W1, W2를 학습해나간다. 

<br/>

#### Code

```python
import torch.nn as nn
# CBOW Model
class CBOWModel(nn.Module):
    def __init__(self, emb_size, emb_dimension):
        super(CBOWModel, self).__init__()
        self.emb_size = emb_size # 단어 집합의 크기
        self.emb_dimension = emb_dimension # 임베딩할 벡터의 차원
        self.u_embeddings = nn.Embedding(2*self.emb_size-1, self.emb_dimension, sparse=True)   #(단어집합크기, 임베딩할 벡터의 차원, sparse)
        self.w_embeddings = nn.Embedding(2*self.emb_size-1, self.emb_dimension, sparse=True)
        self._init_embedding()
    
    def _init_embedding(self):
        int_range = 0.5 / self.emb_dimension
        self.u_embeddings.weight.data.uniform_(-int_range, int_range) # 초기 가중치
        self.w_embeddings.weight.data.uniform_(-int_range, int_range)
        
    def compute_context_matrix(self, u):
        pos_u_emb = []
        for per_Xw in u:
            per_u_emb = self.u_embeddings(torch.LongTensor(per_Xw))   #64비트의 부호 있는 정수 -> LongTensor
            # Xw = Vwi
            Vwi = per_u_emb.data.numpy()
            Vwi_numpy = np.sum(Vwi, axis=0)
            Vwi_list = Vwi_numpy.tolist()
            pos_u_emb.append(Vwi_list)
        pos_u_emb = torch.FloatTensor(pos_u_emb)
        return pos_u_emb
      
    # forward(순전파): 학습을 하면서 답을 출력하는 과정
    def forward(self, pos_u, pos_w, neg_u, neg_w):
        pos_u_emb = self.compute_context_matrix(pos_u)
        pos_w_emb = self.w_embeddings(torch.LongTensor(pos_w))
        neg_u_emb = self.compute_context_matrix(neg_u)
        neg_w_emb = self.w_embeddings(torch.LongTensor(neg_w))
        
        score_1 = torch.mul(pos_u_emb, pos_w_emb).squeeze() # 자녀가 왼쪽 노드
        score_2 = torch.sum(score_1, dim=1)
        score_3 = F.logsigmoid(-1 * score_2)
        neg_score_1 = torch.mul(neg_u_emb, neg_w_emb).squeeze() # 자녀가 오른쪽 노드
        neg_score_2 = torch.sum(neg_score_1, dim=1)
        neg_score_3 = F.logsigmoid(neg_score_2)
        
        loss = torch.sum(score_3) + torch.sum(neg_score_3) # P(w | wi)에 log를 씌웠으므로 합으로 표현해도 무방
        return -1 * loss # 최소화 문제로 변환
    
    # 임베딩 결과 저장
    def save_embedding(self, id2word_dict, file_name):
        embeddings = self.u_embeddings.weight.data.numpy()
        file_output = open(file_name, 'w')
        file_output.write('%d %d\n' % (self.emb_size, self.emb_dimension))
        for id, word in id2word_dict.items():
            e = embeddings[id]
            e = ' '.join(map(lambda x: str(x), e))
            file_output.write('%s %s\n' % (word, e))
```

<br/>

### Skip-gram

CBOW와 반대로 **중심 단어에서 주변 단어를 예측**한다.

<img width="422" alt="Screen Shot 2021-06-02 at 9 49 04 PM" src="https://user-images.githubusercontent.com/46921003/120482715-5a52b500-c3ec-11eb-8c9a-03e579090110.png">

#### Logic

1. 기본적인 구조 및 작동 방식은 CBOW와 유사하다.
2. 단 중심 단어에서 주변 단어를 예측하기 때문에 projection layer에서 입력 벡터들의 평균 값을 구하는 과정이 없다.
3. 따라서 CBOW보다 성능이 뛰어난 것으로 알려져 있다. 

#### Code

```python
# Skip-gram model
class skipgram(nn.Module):
    def __init__(self, vocab_size, embedding_dim):
        super(skipgram, self).__init__()
        self.vocab_size = vocab_size  # vocab_size: 임베딩을 할 단어들의 개수(V)
        self.embedding_dim = embedding_dim  # embedding_dim: 임베딩을 할 벡터의 차원(N)
        # nn.Embedding(): 학습 가능한 임베딩 테이블을 만들기 
        self.v_embeddings = nn.Embedding(vocab_size, embedding_dim, sparse=True)  # W(input vector representation of word w)
        self.init_emb()

    # initialize weights: 처음에는 임의로 설정, 학습 진행되며 업데이트 시킬 부분임.
    # self.v_embeddings.weight.data.uniform_: 균일 분포로 초기화시킴.
    def init_emb(self):
        initrange = 0.5 / self.embedding_dim  # 그냥 임의로 설정한 부분
        self.v_embeddings.weight.data.uniform_(-initrange, initrange)

    # forward(순전파): 학습을 하면서 답을 출력하는 과정
    # squeeze: 1인 차원을 제거한다.
    def forward(self, center_id):
        # center_id: center word의 id
        embed_center = self.v_embeddings(torch.LongTensor([center_id])).squeeze()  # embed_center: hidden layer의 벡터

        return embed_center

    # word vector 저장
    def save_embedding(self, file_name, id2word):
        embeds = self.v_embeddings.weight.data.numpy()  # W
        fo = open(file_name, 'w')
        for idx in range(len(embeds)):
            word = id2word[idx]
            embed = ', '.join(map(str, embeds[idx]))  # 공백을 구분자로 word vector 가지고 오기 ex) 0.5 2.3 0.1
            fo.write(word+' '+embed+'\n')
```
